### PR TITLE
Fix multer vulnerability

### LINKED ISSUE #98: old version of multer is a security risk
As per the code it mentions we are using an old version of multer due to an old bug https://github.com/expressjs/multer/issues/212. This has now been fixed in the never versions of multer.

The old version of multer is reported as a security risk https://snyk.io/test/github/BigstickCarpet/swagger-express-middleware 
middleware_vunerability

### LINKED ISSUE #164: Old version of Multer is causing a high severity vulnerability
We started getting high security vulnerability warnings due to an outdated version of multer. Looks like the package used is over five years old and requires a major version bump resulting in some breaking changes that I wasn't sure how to resolve.
https://github.com/APIDevTools/swagger-express-middleware/blob/c0268efb41f726b2bb57d3b67684812c362a1eaa/package.json#L64

https://www.npmjs.com/advisories/1469

Also looks like this is also fixed now as of `1.0.5` https://github.com/expressjs/multer/issues/212
https://github.com/APIDevTools/swagger-express-middleware/blob/c0268efb41f726b2bb57d3b67684812c362a1eaa/lib/request-parser.js#L36

### PR DESCRIPTION
**Pull request by jphenow**
Fixes #98 

This bumps multer to get rid of a high-risk security vulnerability. It became apparent that multer and a deep amount of it's API surface area leaks out through this library. This means that this PR, as presented, introduces some slight API changes to itself. I did my best to reduce those but this is where I've gotten so far. I'll make the most notable callouts as much as I can.

---

### Storage:

for multer pre-v1 you could option an `inMemory` option to multer to express that you wanted files to be stored in a buffer. They've moved to a storage engine system so that option is now something more like: 

```js
const storage = multer.memoryStorage();
[
  // ...
  middleware.parseRequest({ multipart: { storage } })
 // ...
]
```

### File changes:

Pre v1 multer would store the file with the extension it was uploaded as. This is no longer the case, which throws off some direct sendFile interaction. I wound up needing to update a test helper to explicitly use the File object's mimetype.

There were some slight interface changes to the File object passed to the handler from the middleware. You can see the changes in the tests that illustrate this. Most notably, `buffer` is no longer readily available. This is different depending on the storage class you use. You may receive a buffer using memoryStorage, but not with the default disk storage. It's also worth noting that the library uses `defineProperty` to avoid accidentally leaking anything for the `stream` of data from the file. I believe the stream is the encouraged way to access the raw data. `truncated` is also gone and I believe irrelevant now. There were some new properties added, and using those I was able to backwards-compat add some properties that were expected.

Multer also now explicitly takes note of the index of form-data so `foo[0]=1&foo[2]=30` will now effectively equal: `foo: ["1", undefined, "30"]`.

Note that anonymous files from multer (not explicitly named in swagger fields) are an array, not an object where the field name is the key.

---

By the way I wanted to note that the mix of vscode configurations for this and extensive tests made making this change with any degree of confidence really awesome. Nice work with this library and making it something others can confidently contribute to!
🎊

### CODE REVIEW

**Comment by JamesMessinger** [comment]
Awesome!  Thank you for doing this.  It'll definitely need to be a major version bump, due to API changes, but that's okay.

It's been a long time since I last tried to upgrade Multer, but I believe one of the issues I ran into was the way it handles multiple files in a single request.  I don't recall the details though, sorry.

**Comment by jphenow** [comment]
Agreed - I don't know that it's worth going through the added efforts to make the API match exactly what it was; I'm not even sure how possible it is given the changes. Tests should be green now, going to take off draft and I'll drop some notes on the code for points of interest.

File: lib/mock/index.js [hunk 1]
```diff
@@ -292,6 +292,7 @@ function sendFile (req, res, next, file) {
         return name.toLowerCase() === "content-disposition";
       });
 
+      res.set("Content-Type", file.mimetype);
```
Comment (jphenow) [inline lib/mock/index.js]: files placed on disk are not `/somedir/random-characters`; I think they used to be `/somedir/random-character.jpg` for example. The missing extension is throwing off the auto-mimetype that sendfile (I think provides) so this is a notable change to callout to folks.

I didn't think it was worth somehow renaming the file on disk to keep this behavior.

File: lib/param-parser.js [hunk 2]
```diff
@@ -56,7 +56,12 @@ function parseFormDataParams (req, res, next) {
 
       if (param.type === "file") {
         // Validate the file (min/max size, etc.)
-        req.files[param.name] = parseParameter(param, req.files[param.name], param);
+        let file;
+        if (req.files[param.name] && req.files[param.name][0]) {
+          file = req.files[param.name][0];
```
Comment (jphenow) [inline lib/param-parser.js]: This weird let hand-off is so that we wind up with an undefined in the "right" cases for parseParameter so it validates empty values as expected.

File: lib/request-parser.js [hunk 3]
```diff
@@ -7,10 +7,75 @@ const bodyParser = require("body-parser");
...
 
 // Clean-up the temp directory, even if the app crashes
 tmp.setGracefulCleanup();
 
+/**
+ * Generates a middleware that parses multipart/form-data
+ */
+function generateMultipartFormDataMiddleware (options) {
+  options.allowAll = options.allowAll === undefined ? true : options.allowAll;
+  const uploader = multer(options);
```
Comment (jphenow) [inline lib/request-parser.js]: Traditionally one would pass in the `inMemory` flag; thoughts on making that compatible here? Otherwise folks will need to pull in multer and pass in a storage engine for in-memory. Which is fine, though it seems like a leaky abstraction point possibly?

File: lib/request-parser.js [hunk 4]
```diff
@@ -7,10 +7,75 @@ const bodyParser = require("body-parser");
...
+          fileFields.push({
+            name: param.name,
+            maxCount: 1
+          });
+        }
+      });
+
+      uploader.fields(fileFields)(req, res, (err) => {
+        if (err && err.code === "LIMIT_UNEXPECTED_FILE") {
+          next(); // let request-validator handle
```
Comment (jphenow) [inline lib/request-parser.js]: specifically for making 413 error work correctly in the request validation. I'm open to other approaches.

File: lib/request-parser.js [hunk 5]
```diff
@@ -7,10 +7,75 @@ const bodyParser = require("body-parser");
...
+  };
+}
+
+function multerCompatability (req, res, next) {
+  function standardize (file) {
+    if (file.originalname) {
+      file.extension = path.extname(file.originalname).slice(1);
+    }
+    if (file.filename) {
+      file.name = file.filename;
```
Comment (jphenow) [inline lib/request-parser.js]: I'm personally fine dropping these, it just made some of my initial test changes easier since there was less to fix in the tests if I was able to backfill these values.

File: lib/request-parser.js [hunk 6]
```diff
@@ -7,10 +7,75 @@ const bodyParser = require("body-parser");
...
+    }
+  }
+
+  if (req.files) {
+    if (req.files.length > 0) {
+      req.files.forEach((file) => standardize(file));
```
Comment (jphenow) [inline lib/request-parser.js]: if we allow "any" files, they come in as in array, but we want to standardize as shown above.

File: lib/request-parser.js [hunk 7]
```diff
@@ -7,10 +7,75 @@ const bodyParser = require("body-parser");
...
+    }
+    if (Object.keys(req.files).length > 0) {
+      Object.keys(req.files).forEach((filekey) => {
+        const filearray = Array.from(req.files[filekey]);
```
Comment (jphenow) [inline lib/request-parser.js]: If they are specified fields, they will generally come in as named properties on an object and will sometimes be an array, so if we standardize on looping over an array for standardization we avoid some more complex scenarios.

File: test/specs/mock/delete-collection.spec.js [hunk 8]
```diff
@@ -543,14 +543,14 @@ for (let spec of specs) {
...
+                          destination: res2.body[0].Photo.destination,
+                          filename: res2.body[0].Photo.filename,
                           name: res2.body[0].Photo.name,
                           encoding: "7bit",
                           mimetype: "image/jpeg",
                           path: res2.body[0].Photo.path,
                           extension: "jpg",
                           size: 683709,
-                          truncated: false,
-                          buffer: null
```
Comment (jphenow) [inline test/specs/mock/delete-collection.spec.js]: This file is a good example of the changes. I believe if `memoryStorage` is used you will actually get a buffer here, and possibly `truncated`, though I believe `stream` is the encouraged access point regardless. Also notable that stream is `defineProperty` so it doesn't show up when serialized.

File: lib/mock/index.js (hunk 1 above)
Comment (JamesMessinger) [reply lib/mock/index.js]: Can you add a 4.0.0 entry to [the changelog](https://github.com/APIDevTools/swagger-express-middleware/blob/master/CHANGELOG.md) and call-out any major changes like this?

File: lib/request-parser.js (hunk 3 above)
Comment (JamesMessinger) [reply lib/request-parser.js]: Can't we just check for the `inMemory` flag, and if it's set, then we create a storage engine automatically?

File: lib/request-parser.js (hunk 4 above)
Comment (JamesMessinger) [reply lib/request-parser.js]: This looks good.  Please call-out any behavioral changes in the changelog.

File: lib/request-parser.js (hunk 6 above)
Comment (JamesMessinger) [reply lib/request-parser.js]: agreed

File: test/specs/mock/delete-collection.spec.js (hunk 8 above)
Comment (JamesMessinger) [reply test/specs/mock/delete-collection.spec.js]: This is fine.  But please call it out in the changelog for 4.0.0.

File: test/specs/request-parser.spec.js [hunk 9]
```diff
@@ -676,7 +676,7 @@ for (let spec of specs) {
             expect(req.body).to.deep.equal({
               foo: "bar",
               biz: ["42", "43", "44"],
-              baz: ["B", "C", "A"],
+              baz: ["B", undefined, "C", undefined, undefined, "A"],
```
Comment (JamesMessinger) [inline test/specs/request-parser.spec.js]: This is a weird change.  Is this because of the Multer upgrade?
I'm fine with it, but definitely need to call it out in the changelog for 4.0.0.

File: lib/request-parser.js (hunk 3 above)
Comment (jphenow) [reply lib/request-parser.js]: Yea that was my question basically - if we wanted to do that 🙂 I'll add that.

File: test/specs/request-parser.spec.js (hunk 9 above)
Comment (jphenow) [reply test/specs/request-parser.spec.js]: Yea I thought I had left a comment for it. If you look at the params it basically puts them exactly to the index in the array they are assigned and fills in the blanks with `undefined`. I will note in changelog.

Comment (jphenow) [reply test/specs/request-parser.spec.js]: Looking around more I see form-url-encoded would be ["B", "C", "A"] - I don't know how well I could write up something that would make this compatible but it's odd that the two different content-types would express this slightly differently. Thoughts on what to do there?

File: lib/request-parser.js (hunk 3 above)
Comment (jphenow) [reply lib/request-parser.js]: Added [here](https://github.com/APIDevTools/swagger-express-middleware/pull/165/files#diff-0819ff236dc445648af37b543cd2b958R22-R24)

File: lib/request-parser.js (hunk 5 above)
Comment (jphenow) [reply lib/request-parser.js]: of note I did remove the `name` compat. It seemed silly to support this in perpetuity when it's just a rename. If someone needs it to work as before I think they can write a middleware to handle that on their apps.

**Comment by jphenow** [comment]
Other than the open question about the indexed params noted above, I think this is good for another look! Thanks for taking a look at this stuff with me! Interested in any comments you'd have for Changelog - I erred on the side of over-information.

File: lib/request-parser.js (hunk 5 above)
Comment (JamesMessinger) [reply lib/request-parser.js]: agreed.  Now's the time to make the change, since we're bumping the major version

File: test/specs/request-parser.spec.js (hunk 9 above)
Comment (JamesMessinger) [reply test/specs/request-parser.spec.js]: Yeah, it's weird.  But I think this is an edge-case.  We'll just consider it the new normal and expected behavior, starting with 4.0.0
